state/protocol/kvstore.go

```diff
@@ -101,8 +108,104 @@ type KVStoreReader interface {
 	//    enough that the network is overwhelming likely to finalize at least one
 	//    block with a view in this range
 	GetFinalizationSafetyThreshold() uint64
+
+	// v2
+
+	// GetCadenceComponentVersion returns the Cadence component version.
+	// Error Returns:
+	//   - kvstore.ErrKeyNotSupported if invoked on a KVStore instance before v2.
+	//   - kvstore.ErrKeyNotSet if the key has no value
+	GetCadenceComponentVersion() (MagnitudeOfChangeVersion, error)
+	// GetCadenceComponentVersionUpgrade returns the most recent upgrade for the Cadence Component Version,
+	// if one exists (otherwise returns nil). The upgrade will be returned even if it has already been applied.
+	// Returns nil if invoked on a KVStore instance before v2.
+	GetCadenceComponentVersionUpgrade() *ViewBasedActivator[MagnitudeOfChangeVersion]
+
+	// GetExecutionComponentVersion returns the Execution component version.
+	// Error Returns:
+	//   - kvstore.ErrKeyNotSupported if invoked on a KVStore instance before v2.
+	//   - kvstore.ErrKeyNotSet if the key has no value
+	GetExecutionComponentVersion() (MagnitudeOfChangeVersion, error)
+	// GetExecutionComponentVersionUpgrade returns the most recent upgrade for the Execution Component Version,
+	// if one exists (otherwise returns nil). The upgrade will be returned even if it has already been applied.
+	// Returns nil if invoked on a KVStore instance before v2.
+	GetExecutionComponentVersionUpgrade() *ViewBasedActivator[MagnitudeOfChangeVersion]
+
+	// GetExecutionMeteringParameters returns the Execution metering parameters.
+	// Error Returns:
+	//   - kvstore.ErrKeyNotSupported if invoked on a KVStore instance before v2.
+	//   - kvstore.ErrKeyNotSet if the key has no value
+	GetExecutionMeteringParameters() (ExecutionMeteringParameters, error)
+	// GetExecutionMeteringParametersUpgrade returns the most recent upgrade for the Execution Metering Parameters,
+	// if one exists (otherwise returns nil). The upgrade will be returned even if it has already been applied.
+	// Returns nil if invoked on a KVStore instance before v2.
+	GetExecutionMeteringParametersUpgrade() *ViewBasedActivator[ExecutionMeteringParameters]
+}
+
+// ExecutionMeteringParameters are used to measure resource usage of transactions,
+// which affects fee calculations and transaction/script stopping conditions.
+// TODO should this live in fvm package?
+type ExecutionMeteringParameters struct {
+	// TODO docs
+	ExecutionEffortParameters map[uint]uint64
+	// TODO docs
+	ExecutionMemoryParameters map[uint]uint64
```

Comment:
can we change this to `uint64`? Because `uint` is platform dependent ... and by setting it to `uint64` we avoid ambiguity.

Reply:
@janezpodhostnik Can we use a smaller uint size here for keys, or is `uint64` best?

Reply:
The smallest I would possibly go is `uint16`. But if we are not trying to squeeze every possible byte out of this I would say to go with `uint32`

Reply:
https://github.com/onflow/flow-go/pull/7020/commits/7b3cbe9ce8b3e750de1e4113debab34b5690dc3b